File: packages/contracts/src/envelope.ts
// The response envelope every NAWA endpoint returns, and the error-code
// constants that mirror contracts/errors.py.

export interface Envelope<T = unknown> {
  code: number;
  message: string;
  data: T | null;
}

export const ERROR_CODES = {
  INVALID_FIELDS: 400,
  UNAUTHENTICATED: 401,
  UNAUTHORIZED: 403,
  NOT_FOUND: 404,
  CONFLICT: 409,
  RATE_LIMITED: 429,
  INTERNAL: 500,
} as const;

export type ErrorCode = (typeof ERROR_CODES)[keyof typeof ERROR_CODES];

File: services/api/src/nawa_api/utils/envelope.py
"""Response envelope helpers. Every endpoint returns exactly:

{"code": <int mirrors HTTP status>, "message": <str>, "data": <payload | null>}
"""

from typing import Any


def ok(data: Any) -> dict:
    return {"code": 200, "message": "OK", "data": data}


def created(data: Any) -> dict:
    return {"code": 201, "message": "Created", "data": data}


def accepted(data: Any) -> dict:
    return {"code": 202, "message": "Accepted", "data": data}


def fail(code: int, message: str) -> dict:
    return {"code": code, "message": message, "data": None}
